executor/join.go

```diff
@@ -640,91 +640,101 @@ func (e *HashJoinExec) handleFetchAndBuildHashTablePanic(r interface{}) {
-	if err != nil {
-		e.buildFinished <- errors.Trace(err)
-		close(doneCh)
+	e.buildWorkerWaitGroup.Add(1)
+	// One goroutine fetches chunks from its child and put the chunk into e.buildSideResultCh
+	go util.WithRecovery(func() { e.fetchBuildSideRows(ctx) }, e.handleBuildFetchPanic)
+	// e.concurrency goroutines concurrently fetch chunks from e.buildSideResultCh
+	// and put chunks into a hashtable
+	start := time.Now()
+	defer func() { e.rowContainer.stat.buildTableElapse += time.Since(start) }()
+	for i := uint(0); i < e.concurrency; i++ {
+		e.buildWorkerWaitGroup.Add(1)
+		workID := i
+		go util.WithRecovery(func() { e.buildHashTableForList(workID) }, e.handleBuildWorkerPanic)
 	}
-	// Wait fetchBuildSideRows be finished.
-	// 1. if buildHashTableForList fails
-	// 2. if probeSideResult.NumRows() == 0, fetchProbeSideChunks will not wait for the build side.
-	for range buildSideResultCh {
+	e.buildWorkerWaitGroup.Wait()
+}
+
+func (e *HashJoinExec) handleBuildFetchPanic(r interface{}) {
+	if r != nil {
+		e.buildFinished <- errors.Errorf("%v", r)
 	}
-	// Check whether err is nil to avoid sending redundant error into buildFinished.
-	if err == nil {
-		if err = <-fetchBuildSideRowsOk; err != nil {
-			e.buildFinished <- err
-		}
+	e.buildWorkerWaitGroup.Done()
+}
+
+func (e *HashJoinExec) handleBuildWorkerPanic(r interface{}) {
+	if r != nil {
+		e.buildFinished <- errors.Errorf("%v", r)
 	}
+	e.buildWorkerWaitGroup.Done()
 }
 
-// buildHashTableForList builds hash table from `list`.
-func (e *HashJoinExec) buildHashTableForList(buildSideResultCh <-chan *chunk.Chunk) error {
+// prepareForBuild prepares for each build
+func (e *HashJoinExec) prepareForBuild() {
+	e.buildSideResultCh = make(chan *chunk.Chunk, e.concurrency+1)
+	// set enough size to accommodate errors from each goroutine
+	e.buildFinished = make(chan error, 1+e.concurrency)
 	buildKeyColIdx := make([]int, len(e.buildKeys))
 	for i := range e.buildKeys {
 		buildKeyColIdx[i] = e.buildKeys[i].Index
 	}
-	hCtx := &hashContext{
-		allTypes:  e.buildTypes,
-		keyColIdx: buildKeyColIdx,
+	var hCtx []*hashContext
+	for i := uint(0); i < e.concurrency; i++ {
+		aHCtx := &hashContext{
+			allTypes:  e.buildTypes,
+			keyColIdx: buildKeyColIdx,
+		}
+		hCtx = append(hCtx, aHCtx)
 	}
-	var err error
-	var selected []bool
 	e.rowContainer = newHashRowContainer(e.ctx, int(e.buildSideEstCount), hCtx)
 	e.rowContainer.GetMemTracker().AttachTo(e.memTracker)
 	e.rowContainer.GetMemTracker().SetLabel(buildSideResultLabel)
 	e.rowContainer.GetDiskTracker().AttachTo(e.diskTracker)
 	e.rowContainer.GetDiskTracker().SetLabel(buildSideResultLabel)
 	if config.GetGlobalConfig().OOMUseTmpStorage {
 		actionSpill := e.rowContainer.ActionSpill()
-		failpoint.Inject("testRowContainerSpill", func(val failpoint.Value) {
-			if val.(bool) {
-				actionSpill = e.rowContainer.rowContainer.ActionSpillForTest()
-				defer actionSpill.(*chunk.SpillDiskAction).WaitForTest()
-			}
-		})
```

Comment:
Removing those lines will make some tests unstable.....

Reply:
recoveied it.